importccl: lift computed column check to input converter

pbardea opened the pull request:
The current implementation of checking for validating the number of rows
for computed columns for in non-IMPORT INTO backups is inefficient. This
commit moves the check from being performed on every row to only being
performed once per import.

Fixes #55842.

Release note: None

pbardea committed:
importccl: lift computed column check to input converter

The current implementation of checking for validating the number of rows
for computed columns for in non-IMPORT INTO backups is inefficient. This
commit moves the check from being performed on every row to only being
performed once per import.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/55845)
<!-- Reviewable:end -->

nvanbenschoten commented:
That was fast! I'm curious if you see a way to get rid of https://github.com/cockroachdb/cockroach/pull/55845/files#diff-e56c3f528820bd6c95bc632c127d4ef1b5ae752a8accb667690a5e07c7c9b25eR137 as well. What are the cases where `len(p.expectedColumns) == 0`?

pbardea committed:
importccl: compute expected number of data columns only once

This commit moves the computation of finding the expected number of data
columns to the creation of the input converter, rather than
re-calculating it on every row. The memory footprint of loading all of
the visible columns for every row was noticeable.

Release note (performance improvement): CSV imports should now be
slightly faster.

pbardea commented:
We expect that case when there are no "targetted" columns. In `IMPORT INTO my_table(a, b) ...` (a, b) would be the targetted columns. I believe we would also expect it to be 0 on non-IMPORT INTO flavours.

#55846 is built on top of this, and I think that should do the trick. I took a couple of quick profiles and it looks like that change got rid of VisibleColumns from the Row() profile as expected.

pbardea commented:
TFTR!
bors r=dt
